modules/sc-mesh-secure-deployment/src/2_0/features/PHY/PHY_CRA_scripts/SP_CRA_v7.py

```diff
@@ -0,0 +1,214 @@
+        except Exception as e:
+            logging.error("Error during server setup: %s", e)
+
+        # Client setup
+        try:
+            DataClient = loadmat('ACF_Table.mat')
+            self.acf_client = np.array(DataClient['y_C']).transpose()
+            logging.info("Client setup completed successfully")
+        except Exception as e:
+            logging.error("Error during client setup: %s", e)
+        
+        # Clear the server log at the start of each session
+        if os.path.exists("server_log.txt"):
+            os.remove("server_log.txt")
+
+        # Create an event to signal when 5 minutes have passed
+        self.stop_event = threading.Event()
+
+        # Start server and client functionalities
+        try:
+            self.server_thread = threading.Thread(target=self.server_start)
+            self.server_thread.start()
+            logging.info("Server thread started")
+            time.sleep(2)
+            self.listen_thread = threading.Thread(target=self.listen_for_broadcast)
+            self.listen_thread.start()
+            logging.info("Broadcast listening thread started")
+        except Exception as e:
+            logging.error("Error during server/client threads initialization: %s", e)
+
+        # Start a timer to stop the server and broadcasting after 5 minutes
+        timer = threading.Timer(60, self.stop_event.set)
+        timer.start()
+        logging.info("Server shutdown timer started")
+
+    # SERVER FUNCTIONS
+    def log_authentication(self, node_ip, mac_address, result):
+        timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
+        log_entry = f"{timestamp}\t{node_ip}\t{mac_address}\t{result}\n"
+        with open("server_log.txt", "a") as log_file:
+            log_file.write(log_entry)
+
+    def display_table(self):
+        logging.info("Displaying authentication table")
+        print("+---------------------+---------------+-------------------+---------------------------+")
+        print("|        Time         |     Node IP   |    MAC Address    |   Authentication Result   |")
+        print("+---------------------+---------------+-------------------+---------------------------+")
+        with open("server_log.txt", "r") as log_file:
+            for line in log_file:
+                timestamp, node_ip, mac_address, result = line.strip().split("\t")
+                formatted_line = f"| {timestamp:<19} | {node_ip:<12} | {mac_address:<15} | {result:<23} |"
+                print(formatted_line)
+        print("+---------------------+---------------+-------------------+---------------------------+")
+
+    def handle_client(self, conn, addr):
+        logging.info(f"Connection request received from {addr}")
+        print(f"Connection request received from {addr}")
+        try:
+            index = random.randint(0, len(self.acf) - 1)
+            acf_tx = pickle.dumps(self.acf[index])
+            conn.send(acf_tx)
```

Comment:
send() vs sendall() also here?

Reply:
Here, the server just send the auto-correlation value corresponding to a index, which is very small data.

Reply:
https://docs.python.org/3/library/socket.html#socket.socket.send

8<-----8<-----8<-----8<-----8<-----8<-----
Applications are responsible for checking that all data has been sent; if only some of the data was transmitted, the application needs to attempt delivery of the remaining data. For further information on this topic, consult the [Socket Programming HOWTO](https://docs.python.org/3/howto/sockets.html#socket-howto).
8<-----8<-----8<-----8<-----8<-----8<-----